src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -160,17 +155,20 @@ public CompletionStage<UStatus> unsubscribe(UUri topic, UListener listener, Call
 
         final UnsubscribeRequest unsubscribeRequest = UnsubscribeRequest.newBuilder().setTopic(topic).build();
         
-        return RpcMapper.mapResponseToResult(
+        CompletionStage<RpcResult<UnsubscribeResponse>> result = RpcMapper.mapResponseToResult(
             // Send the unsubscribe request
-            rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options), 
-            UnsubscribeResponse.class)
-            
+            rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options),     
+                UnsubscribeResponse.class)
             // Then unregister the listener
-            .thenCompose(response -> {
+            .whenComplete((response, ex) -> mHandlers.remove(topic));
+
+        // Create a CompletionStage that will handle the response from the unsubscribe request
+        return transport.unregisterListener(topic, listener)
+            .thenCombine(result, (status, response) -> {
                 if (response.isSuccess()) {
-                    return transport.unregisterListener(topic, listener);
+                    return UStatus.newBuilder().setCode(UCode.OK).build();
                 }
-                return CompletableFuture.completedFuture(response.failureValue());
+                return response.failureValue();
```

Comment:
In this case, we have (logically) unregistered using USubscription but have failed to unregister the listener from UTransport, right? If so, then a IMHO the JavaDoc should be explicit about this fact and what it means for the client ...

Reply:
done